test/e2e/kafka_source_reconciler_test.go

```diff
@@ -0,0 +1,131 @@
+//+build e2e
+
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	testlib "knative.dev/eventing/test/lib"
+	"knative.dev/eventing/test/lib/resources"
+	"knative.dev/pkg/apis"
+	pkgTest "knative.dev/pkg/test"
+
+	sourcesv1beta1 "knative.dev/eventing-contrib/kafka/source/pkg/apis/sources/v1beta1"
+	"knative.dev/eventing-contrib/test"
+	"knative.dev/eventing-contrib/test/e2e/helpers"
+	contribtestlib "knative.dev/eventing-contrib/test/lib"
+	contribresources "knative.dev/eventing-contrib/test/lib/resources"
+)
+
+const (
+	rtKafkaSourceName    = "e2e-rt-kafka-source"
+	rtChannelName        = "e2e-rt-channel"
+	rtKafkaConsumerGroup = "e2e-rt-cg"
+	rtKafkaTopicName     = "e2e-rt-topic"
+)
+
+//TestKafkaSourceReconciler tests various kafka source reconciler statuses
+//RT is short for reconciler test
+func TestKafkaSourceReconciler(t *testing.T) {
+	client := testlib.Setup(t, true)
+	defer testlib.TearDown(client)
+
+	for _, test := range []struct {
+		name             string
+		action           func(c *testlib.Client)
+		expectedStatuses sets.String
+		wantRADepCount   int
+	}{{
+		"create_kafka_source",
+		createKafkaSourceWithSinkMissing,
+		sets.NewString("NotFound"),
+		0,
+	}, {
+		"create_sink",
+		createChannel,
```

Comment:
I don't get why you're using the kafka channel here :smile:, if you need to create a "sink", can you use a simpler one (like the test images to receive events)?

Reply:
I wanted to use the simplest one like you have mentioned, but if am not wrong, if the sink is ["v1/Service"](https://github.com/knative/pkg/blob/3ff7bcc4225975f76b6cc294a7ca9c5db32de222/resolver/addressable_resolver.go#L154), it is not validated to see if its available or not. Reconciles to true always. Next option was to use ksvc, but i dint see any ksvc used as sink in any of the eventing tests nor the knative serving components getting installed. Thats why used a channel. Willing to change to simpler ones if there are other options.

Reply:
i think you need an addressable, so maybe using a kube service with a pod behind the hood is fine? I would love as much as possible to have this "dependency" between the kafka source tests and kafka channel

Reply:
but with kube service as i mentioned above , its not reporting the right status on the kafks source. It always reconciles to true whether its there are not. I initially tried with kube service only.

Reply:
ok so then, more than creating a kafka channel, can you create an imc channel and clearly state you need it for that purpose?

Reply:
> if the sink is "v1/Service", it is not validated to see if its available or not.

Oh, ok, this is the reason that I wasn't able to reproduce the issue. I was trying with a Kube service

Reply:
used imc initially, the tests failed, then switched to kafka channel. Tried one more time with imc, got the below logs in tests 
```
=== RUN   TestKafkaSourceReconciler/create_sink
=== CONT  TestKafkaSourceReconciler
    creation.go:77: Creating channel &TypeMeta{Kind:InMemoryChannel,APIVersion:messaging.knative.dev/v1beta1,}-e2e-rt-channel
    creation.go:85: Failed to create "InMemoryChannel" "e2e-rt-channel": the server could not find the requested resource
    creation.go:90: Failed to create "InMemoryChannel" "e2e-rt-channel": the server could not find the requested resource
=== CONT  TestKafkaSourceReconciler/create_sink
    testing.go:964: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
=== CONT  TestKafkaSourceReconciler
    test_runner.go:195: EVENT: {{ } {e2e-rt-kafka-source.163333e31993bfcc  test-kafka-source-reconciler-wbvf2 /api/v1/namespaces/test-kafka-source-reconciler-wbvf2/events/e2e-rt-kafka-source.163333e31993bfcc 278249f1-af24-4582-93f9-4811b4e4e239 1172 0 2020-09-09 19:22:23 +0000 UTC <nil> <nil> map[] map[] [] []  []} {KafkaSource test-kafka-source-reconciler-wbvf2 e2e-rt-kafka-source be8ceb0b-957b-4e18-813c-5c40244dbdd3 sources.knative.dev/v1beta1 19196 } InternalError getting sink URI: failed to get lister for messaging.knative.dev/v1beta1, Resource=inmemorychannels: inmemorychannels.messaging.knative.dev is forbidden: User "system:serviceaccount:knative-sources:kafka-controller-manager" cannot list resource "inmemorychannels" in API group "messaging.knative.dev" at the cluster scope {kafkasource-controller } 2020-09-09 19:22:23 +0000 UTC 2020-09-09 19:22:25 +0000 UTC 6 Warning 0001-01-01 00:00:00 +0000 UTC nil  nil 
```

Reply:
The first 2 errors (failed to create) are weird, i don't get what's the problem there tbh, double check all the code you use to start the imc (in fact, in testlib.Client there should be a method to create a imc channel (it's from eventing/test/lib), you don't need to develop it by yourself.

The last error is clearly an issue of cluster roles, seems like you need some specific cluster role to check if the addressable is ready. @matzew any thoughts on that?

Reply:
imc was not set up as part of e2e set up, added those in the e2e scripts and test passes. @slinkydeveloper @matzew